### ISSUE TITLE
`HTTPMethod` should not be an enum

### ISSUE THREAD

**Issue by helje5** [apple/swift-nio#146]
### Expected behavior
No API breakage if an HTTP method is added to the enum.

### Actual behavior
API breakage if an HTTP method is added to the enum.

### SwiftNIO version/commit hash
1.1

As long as Swift doesn't provide an `open enum` construct (which I think was in discussion, don't know the state of it), it is not wise to use a (closed) enum for constants which are unstable (in an API which is supposed to be stable). In this case, HTTP methods are added once in a while by new RFCs (I myself added MKCALENDAR to http_parser).

This was also discussed a lot in the HTTP WG effort, where we finally ended up with the common iOS solution: https://github.com/swift-server/http/blob/develop/Sources/HTTP/HTTPMethod.swift#L10 (I'm not sure this is the best solution, probably not, but it at least provides better API stability).

In short: Adding a new case to HTTPMethod currently requires a major SemVer bump. That is non-ideal as the addition of a HTTP methods is otherwise not a big event / breaking change ;-)

**Comment by Lukasa** [comment]
Is there a reason the `.RAW` escape hatch is insufficient here?

**Comment by helje5** [comment]
Well, first it doesn't change the issue (you cannot add a case w/o a major bump), and second it is ugly :-)

Let's assume we add `BATCH` (which we will), do you want all code to look like
```swift
switch method {
  case .GET: ...
  case .POST: ...
  case .RAW("BATCH"): ...
}
```
Booo! No! ;-)

Now, this may be OK for a sub minor hotfix, but you really want a new .BATCH case and it is kinda stupid to wait for a major release (which is usually major for other reasons ...) to address that. Just because you use a closed type for an open construct outside of your control.

The extreme case of this is: Why don't we use an enum for headers then? ;-)

**Comment by Lukasa** [comment]
I am very much on the fence. I quite like the way the enum solution affords compactness in the vast majority of common cases, but I've also frequently wanted to discourage users from assuming that HTTP verbs are necessarily particularly well known or well formed. So I'd be open to making this kind of change in a 2.0 or 3.0 release.

**Comment by helje5** [comment]
Not sure what you mean by `compactness`. On the user side it is the same and on the implementation side too.

Closed enum:

```swift
public enum HTTPMethod: Equatable {
  case GET
  case PUT
  case BATCH

  public static func ==(lhs: HTTPMethod, rhs: HTTPMethod) -> Bool {
    switch (lhs, rhs) {
      case (.GET,   .GET):   return true
      case (.PUT,   .PUT):   return true
      case (.BATCH, .BATCH): return true
      default:               return false
    }
  }
}
```

User:
```swift
let closed : HTTPMethod = .GET

switch closed {
  case .GET: print("retrieve")
  case .PUT: print("store")
  default:   print("something else")
}
```

Open "Enum":
```swift
public struct OHTTPMethod: Equatable {
  
  let value : Int
  static let GET   = OHTTPMethod(value: 1)
  static let PUT   = OHTTPMethod(value: 2)
  static let BATCH = OHTTPMethod(value: 3)

  public static func ==(lhs: OHTTPMethod, rhs: OHTTPMethod) -> Bool {
    return lhs.value == rhs.value
  }
}
```

User:
```swift
let openMethod : OHTTPMethod = .GET

switch openMethod {
  case .GET: print("retrieve")
  case .PUT: print("store")
  default:   print("something I don't know")
}
```

That would have to wait for 2.0 of course, SemVer and all that ;->

**Comment by Lukasa** [comment]
I mean compactness in terms of memory layout. :smile:

**Comment by helje5** [comment]
Well, it isn't more compact than the above and probably need to be 64bit aligned in any case (so a String pointer would actually be kinda OK too).

**Comment by weissi** [comment]
tagged this 2.0, @helje5 / @Lukasa any new insights into this one?

**Comment by Lukasa** [comment]
I continue to not think this is a very big deal, but I'm definitely not opposed to changing to an alternative representation.

**Comment by helje5** [comment]
As long as there is no „open enum“ (I thought they might be coming w/ 5?) the struct is still the way to, IMO.

**Comment by Lukasa** [comment]
Open enums are stdlib-and-overlay-only in Swift 5, so not available for SwiftNIO.

**Comment by weissi** [comment]
what do folks like @airspeedswift / @moiseev recommend here?

**Comment by weissi** [comment]
Despite the fact that we've got open enums now, I agree that `HTTPMethod` shouldn't be an enum. I think a `struct` with `static let`s is the right answer here but ofc NIO 2.x can't change the existing type.

**Comment by helje5** [comment]
What's the advantage of a struct over an open enum?

**Comment by weissi** [comment]
> What's the advantage of a struct over an open enum?

For enums, we'll always need this annoying `.raw("WHATEVER")` case. But then you have a representational discrepancy for verbs that are actually standard. So `.raw("GET")` should be the same as `.GET` and you can implement `==` to do that. However if somebody would then do

```
guard case .GET = httpRequest.method else {
   throw Abort("only GETs supported")
}
```

the `.raw("GET")` would fail. Yes, you could argue that users should write `guard .GET == httpRequest.method` but that's also less idiomatic.

So the crux with "completely open" `enum`s that have a `.raw(...)` case is that you don't get to turn `.raw("GET")` into `.GET`. A struct doesn't have this problem.

**Comment by Lukasa** [comment]
Arguably this entire issue is moot with swift-http-types now slowly replacing this API. The most likely semver major is actually just RMing this type altogether.
